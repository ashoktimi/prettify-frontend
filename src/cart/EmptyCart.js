import React, { useState, useEffect } from "react";
import { addToCart } from "./cartSlice";
import PrettifyApi from "../api/api";
import LoadingSpinner from "../helpers/LoadingSpinner";
import { useDispatch } from "react-redux";
import { Stack } from "react-bootstrap";

const EmptyCart = () => {
    const [products, setProducts] = useState(null);
    const dispatch = useDispatch();
    useEffect(function getProductsOnMount() {
        search();
      }, []);

    async function search() {
        let products = await PrettifyApi.getRecProducts();
        setProducts(products.products);
    }
  const handleAddToCart = (product) => {
    dispatch(addToCart(product));
  };
  
  if (!products) return <LoadingSpinner />;
  return (
  <>
    <p>Your cart is empty</p>
      <p>Why not check out these recommended products?</p>
        <Stack gap={3}>
        {products.map(item => (
          <Stack gap={2}>
            <Stack direction="horizontal">
            <img
              src={item.image_link}    
              style={{ width: "125px", height: "115px", objectFit: "cover" }}
              alt={''}
            />
              {item.name}{" "}${(item.price)}
            </Stack> 
            
              <Stack direction="vertical" >         
  
                <div style={{display: "flex", justifyContent: 'center'}}>
                  <button onClick={() => handleAddToCart(item)} style={{border:'none', backgroundColor:'white',  marginRight: '15px'}}> add to cart </button>
                 
              </div>   
              </Stack>
          </Stack>
  
        ))}
      </Stack>
      </>
  );
};

export default EmptyCart;
