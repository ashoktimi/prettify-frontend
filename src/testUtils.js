import React from "react";
import UserContext from "./auth/userContext";


const demoUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
  photo_url: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, handleToggleOffcanvas = () => false }) => (
    <UserContext.Provider value={{ currentUser, handleToggleOffcanvas }}>
      {children}
    </UserContext.Provider>
);

function setup(){
  window.matchMedia = jest.fn().mockImplementation(query => {
    return {
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    };
  });
}

function teardown(){
   window.matchMedia.mockClear();
   delete window.matchMedia;
}

function commonBeforeEach() {
  setup();
}

function commonAfterEach() {
  teardown();
}

export { UserProvider, commonBeforeEach, commonAfterEach };