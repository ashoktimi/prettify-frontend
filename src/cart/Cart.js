import { useEffect, useContext } from "react";
import { useDispatch, useSelector } from "react-redux";
import UserContext from "../auth/userContext";
import EmptyCart from "./EmptyCart";

import { RiDeleteBinLine } from 'react-icons/ri';
import {
  addToCart,
  clearCart,
  decreaseCart,
  getTotals,
  removeFromCart,
} from "./cartSlice";
import { Offcanvas, Stack, Button } from "react-bootstrap";

const Cart = () => {
  const cart = useSelector((state) => state.cart);
  const dispatch = useDispatch();
  const { showOffcanvas, handleToggleOffcanvas } = useContext(UserContext);


  useEffect(() => {
    dispatch(getTotals());
  }, [cart, dispatch]);

  const handleAddToCart = (product) => {
    dispatch(addToCart(product));
  };
  const handleDecreaseCart = (product) => {
    dispatch(decreaseCart(product));
  };
  const handleRemoveFromCart = (product) => {
    dispatch(removeFromCart(product));
  };
  const handleClearCart = () => {
    dispatch(clearCart());
  };

  console.log((cart.cartItems).length)

  return (
    <>
    <Offcanvas show={showOffcanvas} onHide={handleToggleOffcanvas} placement="end">
      <Offcanvas.Header closeButton>
        <Offcanvas.Title style={{left:0, right:0, margin:'auto'}}>Shopping Cart</Offcanvas.Title>
      </Offcanvas.Header>
      <Offcanvas.Body>
    {(cart.cartItems).length?
    <>
    <Stack gap={3}>
      {cart.cartItems.map(item => (
        <Stack gap={2}>
          <Stack direction="horizontal">
          <img
            src={item.image_link}    
            style={{ width: "125px", height: "115px", objectFit: "cover" }}
            alt={''}
          />
            {item.name}{" "}${(item.price)}
            {item.cartQuantity > 1 && (
                <span className="text-muted" style={{ fontSize: ".85rem", letterSpacing:'30px'}}>
                  x{item.cartQuantity }
                </span>
            )}
            {item.cartQuantity > 1?
            <div style={{display: "flex", justifyContent: 'flex-end'}}> 
                ${(item.price * item.cartQuantity)}
            </div> : null
            }
          </Stack>        
          
            <Stack direction="vertical" >         

              <div style={{display: "flex", justifyContent: 'center'}}>
                <button onClick={() => handleDecreaseCart(item)} style={{border:'none', backgroundColor:'white', marginRight: '15px'}}>
                  -
                </button>
                <div style={{ marginRight: '15px'}}>{item.cartQuantity}</div>
                <button onClick={() => handleAddToCart(item)} style={{border:'none', backgroundColor:'white',  marginRight: '15px'}}> + </button>
               
              <Button
                variant="light"
                onClick={() => handleRemoveFromCart(item)}
                style={{backgroundColor: 'white'}}               
              >
                <RiDeleteBinLine />
            </Button>
            </div>   
            </Stack>
        </Stack>

      ))}
      <Button 
        className="ms-auto"
        variant="outline-danger"
        size="sm" 
        onClick={()=> handleClearCart()} 
        style={{width:'fit-content'}}  
       
      >
        Clear Cart
      </Button>
    </Stack>


      <div className="subtotal"style={{ display: 'flex', justifyContent: 'space-between', marginTop: '1rem'}}>
        <span style={{marginLeft: '15px'}}>Subtotal : </span>
        <span className="amount">${cart.cartTotalAmount}</span>     
      </div>
      <Button href="/checkout" variant="warning" style={{ margin: '5rem'}}>Check Out</Button>

      </>
      :
      <EmptyCart />
      }
      </Offcanvas.Body>
    </Offcanvas>
    </>
  )

};

export default Cart;