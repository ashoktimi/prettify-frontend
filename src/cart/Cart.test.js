import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { UserProvider, commonBeforeEach, commonAfterEach } from "../testUtils";
import { configureStore } from '@reduxjs/toolkit';
import cartReducer from './cartSlice';
import Cart from './Cart';
import EmptyCart from "./EmptyCart";


beforeEach(commonBeforeEach)
afterEach(commonAfterEach)

describe('Cart', () => {
 const store = configureStore({
   reducer: {
     cart: cartReducer
   }
 });

 test("renders without crashing", function () {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <UserProvider>
            <Cart />
          </UserProvider>
        </Provider>
      </MemoryRouter>
     );
  });
  
  test("matches snapshot", function () {
    const { asFragment } = 
     render(
      <MemoryRouter>
        <Provider store={store}>
          <UserProvider>
            <Cart />
          </UserProvider>
        </Provider>
      </MemoryRouter>
     );
    expect(asFragment()).toMatchSnapshot();
  });

  test('EmptyCart component renders correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
           <EmptyCart />       
        </Provider>    
      </MemoryRouter> 
    );
    expect(container).toMatchSnapshot();
  });
})