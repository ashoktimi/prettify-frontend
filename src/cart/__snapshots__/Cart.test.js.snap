// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart EmptyCart component renders correctly 1`] = `
<div>
  <div
    class="LoadingSpinner"
  >
    Loading ...
  </div>
</div>
`;

exports[`Cart matches snapshot 1`] = `<DocumentFragment />`;
